xapian-core/tests/harness/backendmanager_honey.cc

```diff
@@ -73,12 +73,62 @@ BackendManagerHoney::do_get_database_path(const vector<string> & files)
     return dbpath;
 }
 
+Xapian::WritableDatabase
+BackendManagerHoney::get_generated_database(const string& name)
+{
+    // creating generated database inside glass cache
+    // to prevent a valid glass db inside honey cache
+    // if testsuite was interrupted
+    string db_path = ".glass/" + name;
+
+    rm_rf(db_path);
+
+    auto flags = Xapian::DB_CREATE|Xapian::DB_BACKEND_GLASS;
+    Xapian::WritableDatabase wdb(db_path, flags);
+
+    return wdb;
+}
+
+void
+BackendManagerHoney::finalise_generated_database(const string& name)
+{
+    // converting a glass backend to honey
+
+    // path to the generated glass db
+    string glass_db_path = ".glass/" + name;
+
+    // path to honey tmpfile
+    string tmpfile = CACHE_DIRECTORY "/" + name;
+    tmpfile += ".tmp";
+
+    // path to final honey db
+    string path = CACHE_DIRECTORY "/" + name;
+
+    Xapian::WritableDatabase wdb(glass_db_path);
+    wdb.compact(tmpfile, Xapian::DB_BACKEND_HONEY);
+    wdb.close();
+
+    rename(tmpfile.c_str(), path.c_str());
+}
+
 Xapian::WritableDatabase
 BackendManagerHoney::get_writable_database(const string &, const string &)
 {
     throw Xapian::UnimplementedError("Honey databases don't support writing");
 }
 
+string
+BackendManagerHoney::get_generated_database_path(const string& name)
+{
+    // the generated database would go inside the glass cache
+    // if it does not exist yet
+    string path = CACHE_DIRECTORY "/" + name;
+    if (path_exists(path)) {
+        return path;
+    }
+    return ".glass/" + name;
```

Comment:
That's quite cunning, but we shouldn't need this peculiar alternating return value if we delegate the creation of the glass database to `BackendManagerGlass` - we can just pass through the generator function to that and it'll handle creating the glass database and calling the generator function on it (if it doesn't already exist).  Then we just need to convert that.